File: src/_5_polish/java/Calculator.java
package _5_polish.java;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import static _5_polish.java.ArithmeticStack.*;

public class Calculator {

    static private Map<String, Operation> operationMap = new HashMap<>();
    static {
        operationMap.put("+", Operation.PLUS);
        operationMap.put("-", Operation.MINUS);
        operationMap.put("*", Operation.TIMES);
        operationMap.put("/", Operation.DIV);
    }

    static public double calc(String expr) {
        final ArithmeticStack stack = new ArithmeticStack();
        for (String arg: expr.split(" ")) {
            Operation op = operationMap.get(arg);
            if (op==null) {
                double x = Double.parseDouble(arg);
                stack.push(x);
            }
            else {
                stack.execute(op);
            }
        }
        return stack.top();
    }

    static public double calcFunctional(String expr) {
        final ArithmeticStack stack = new ArithmeticStack();
        Arrays.asList(expr.split(" ")).forEach(s -> {
            Operation op = operationMap.get(s);
            if (op==null) {
                double x = Double.parseDouble(s);
                stack.push(x);
            }
            else {
                stack.execute(op);
            }
        });
        return stack.top();
    }
}

File: src/_5_polish/kotlin/Calculator.kt
package _5_polish.kotlin

import java.util.*

// Singleton
object Calculator {
    private val operationMap = mapOf<String, (Double, Double) -> Double>(
            "+" to { x, y -> x + y },
            "-" to { x, y -> y - x },
            "*" to { x, y -> x * y },
            "/" to { x, y -> x / y }
    )

    fun Stack<Double>.execute(op: (Double, Double) -> Double) = push(op(pop(), pop()))

    fun calc(expr: String): Double {
        val stack = Stack<Double>()
        for (arg in expr.split(" ")) {
            val op = operationMap[arg]
            if (op != null) {
                stack.execute(op)
            }
            else {
                stack.push(arg.toDouble())
            }
        }
        return stack.pop()
    }

    fun calcFunctional(expr: String): Double {
        val stack = Stack<Double>()
        // Extension: for-each
        expr.split(" ").forEach {
            // it = single argument
            // let: receiver --> argument                     // elvis, toDouble()
            operationMap[it]?.let { op -> stack.execute(op) } ?: stack.push( it.toDouble() )
        }
        return stack.pop()
    }
}